In [26]:
import pandas as pd
import os

In [27]:
folder="/home/durga/Desktop/DURGA/E2S1/DS/Course/NIFTY-Stock-Dashboard/dataset"
files=os.listdir(folder)
print("total files",len(files))
print("files",files)

total files 52
files ['NIFTY50_all.csv', 'UPL.csv', 'TECHM.csv', 'EICHERMOT.csv', 'GRASIM.csv', 'IOC.csv', 'HDFC.csv', 'stock_metadata.csv', 'INDUSINDBK.csv', 'COALINDIA.csv', 'JSWSTEEL.csv', 'SBIN.csv', 'INFRATEL.csv', 'ITC.csv', 'TITAN.csv', 'RELIANCE.csv', 'POWERGRID.csv', 'NESTLEIND.csv', 'SHREECEM.csv', 'ASIANPAINT.csv', 'ONGC.csv', 'NTPC.csv', 'HINDUNILVR.csv', 'TCS.csv', 'BAJAJ-AUTO.csv', 'TATASTEEL.csv', 'AXISBANK.csv', 'BRITANNIA.csv', 'HCLTECH.csv', 'ULTRACEMCO.csv', 'BPCL.csv', 'BHARTIARTL.csv', 'MM.csv', 'BAJFINANCE.csv', 'VEDL.csv', 'SUNPHARMA.csv', 'ZEEL.csv', 'HINDALCO.csv', 'CIPLA.csv', 'KOTAKBANK.csv', 'BAJAJFINSV.csv', 'ADANIPORTS.csv', 'LT.csv', 'WIPRO.csv', 'TATAMOTORS.csv', 'HEROMOTOCO.csv', 'MARUTI.csv', 'ICICIBANK.csv', 'DRREDDY.csv', 'HDFCBANK.csv', 'INFY.csv', 'GAIL.csv']


In [28]:
df=pd.read_csv(os.path.join(folder,files[0]))
df.head()

,Date,Symbol,Series,Prev Close,Open,High,Low,Last,Close,VWAP,Volume,Turnover,Trades,Deliverable Volume,%Deliverble
0,2007-11-27,MUNDRAPORT,EQ,440.00,770.00,1050.00,770.0,959.0,962.90,984.72,27294366,2.687719e+15,NaN,9859619.0,0.3612
1,2007-11-28,MUNDRAPORT,EQ,962.90,984.00,990.00,874.0,885.0,893.90,941.38,4581338,4.312765e+14,NaN,1453278.0,0.3172
2,2007-11-29,MUNDRAPORT,EQ,893.90,909.00,914.75,841.0,887.0,884.20,888.09,5124121,4.550658e+14,NaN,1069678.0,0.2088
3,2007-11-30,MUNDRAPORT,EQ,884.20,890.00,958.00,890.0,929.0,921.55,929.17,4609762,4.283257e+14,NaN,1260913.0,0.2735
4,2007-12-03,MUNDRAPORT,EQ,921.55,939.75,995.00,922.0,980.0,969.30,965.65,2977470,2.875200e+14,NaN,816123.0,0.2741


In [29]:
df_reliance = df[df["Symbol"]=="RELIANCE"].copy()

In [30]:
df_reliance

,Date,Symbol,Series,Prev Close,Open,High,Low,Last,Close,VWAP,Volume,Turnover,Trades,Deliverable Volume,%Deliverble
166253,2000-01-03,RELIANCE,EQ,233.05,237.50,251.70,237.50,251.70,251.70,249.37,4456424,1.111319e+14,NaN,NaN,NaN
166254,2000-01-04,RELIANCE,EQ,251.70,258.40,271.85,251.30,271.85,271.85,263.52,9487878,2.500222e+14,NaN,NaN,NaN
166255,2000-01-05,RELIANCE,EQ,271.85,256.65,287.90,256.65,286.75,282.50,274.79,26833684,7.373697e+14,NaN,NaN,NaN
166256,2000-01-06,RELIANCE,EQ,282.50,289.00,300.70,289.00,293.50,294.35,295.45,15682286,4.633254e+14,NaN,NaN,NaN
166257,2000-01-07,RELIANCE,EQ,294.35,295.00,317.90,293.00,314.50,314.55,308.91,19870977,6.138388e+14,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
171554,2021-04-26,RELIANCE,EQ,1904.35,1920.00,1962.00,1911.50,1938.00,1937.85,1941.32,9620785,1.867699e+15,259137.0,4276703.0,0.4445
171555,2021-04-27,RELIANCE,EQ,1937.85,1940.00,1997.20,1938.25,1990.00,1988.65,1978.64,9226547,1.825602e+15,291197.0,3772144.0,0.4088
171556,2021-04-28,RELIANCE,EQ,1988.65,1997.85,2008.00,1980.15,1993.15,1997.30,1997.60,7902002,1.578508e+15,247331.0,3921560.0,0.4963
171557,2021-04-29,RELIANCE,EQ,1997.30,2022.90,2044.50,2007.30,2020.00,2024.05,2024.21,8035915,1.626634e+15,213153.0,2834103.0,0.3527


In [31]:
df_reliance.isnull().sum()

Date                     0
Symbol                   0
Series                   0
Prev Close               0
Open                     0
High                     0
Low                      0
Last                     0
Close                    0
VWAP                     0
Volume                   0
Turnover                 0
Trades                2850
Deliverable Volume     514
%Deliverble            514
dtype: int64

In [32]:
df_reliance.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5306 entries, 166253 to 171558
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Date                5306 non-null   object 
 1   Symbol              5306 non-null   object 
 2   Series              5306 non-null   object 
 3   Prev Close          5306 non-null   float64
 4   Open                5306 non-null   float64
 5   High                5306 non-null   float64
 6   Low                 5306 non-null   float64
 7   Last                5306 non-null   float64
 8   Close               5306 non-null   float64
 9   VWAP                5306 non-null   float64
 10  Volume              5306 non-null   int64  
 11  Turnover            5306 non-null   float64
 12  Trades              2456 non-null   float64
 13  Deliverable Volume  4792 non-null   float64
 14  %Deliverble         4792 non-null   float64
dtypes: float64(11), int64(1), object(3)
memory usage: 663

In [33]:
df_reliance.isnull().sum()

Date                     0
Symbol                   0
Series                   0
Prev Close               0
Open                     0
High                     0
Low                      0
Last                     0
Close                    0
VWAP                     0
Volume                   0
Turnover                 0
Trades                2850
Deliverable Volume     514
%Deliverble            514
dtype: int64

In [34]:
df_reliance["Date"]=pd.to_datetime(df_reliance["Date"])

In [35]:
df_reliance.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5306 entries, 166253 to 171558
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Date                5306 non-null   datetime64[ns]
 1   Symbol              5306 non-null   object        
 2   Series              5306 non-null   object        
 3   Prev Close          5306 non-null   float64       
 4   Open                5306 non-null   float64       
 5   High                5306 non-null   float64       
 6   Low                 5306 non-null   float64       
 7   Last                5306 non-null   float64       
 8   Close               5306 non-null   float64       
 9   VWAP                5306 non-null   float64       
 10  Volume              5306 non-null   int64         
 11  Turnover            5306 non-null   float64       
 12  Trades              2456 non-null   float64       
 13  Deliverable Volume  4792 non-null   float64   

In [36]:
df_reliance.fillna(df.mean(numeric_only=True),inplace=True)

In [37]:
df_reliance["target"]=df_reliance["Close"].shift(-1)

In [38]:
df_reliance

,Date,Symbol,Series,Prev Close,Open,High,Low,Last,Close,VWAP,Volume,Turnover,Trades,Deliverable Volume,%Deliverble,target
166253,2000-01-03,RELIANCE,EQ,233.05,237.50,251.70,237.50,251.70,251.70,249.37,4456424,1.111319e+14,61964.270982,1.315098e+06,0.502997,271.85
166254,2000-01-04,RELIANCE,EQ,251.70,258.40,271.85,251.30,271.85,271.85,263.52,9487878,2.500222e+14,61964.270982,1.315098e+06,0.502997,282.50
166255,2000-01-05,RELIANCE,EQ,271.85,256.65,287.90,256.65,286.75,282.50,274.79,26833684,7.373697e+14,61964.270982,1.315098e+06,0.502997,294.35
166256,2000-01-06,RELIANCE,EQ,282.50,289.00,300.70,289.00,293.50,294.35,295.45,15682286,4.633254e+14,61964.270982,1.315098e+06,0.502997,314.55
166257,2000-01-07,RELIANCE,EQ,294.35,295.00,317.90,293.00,314.50,314.55,308.91,19870977,6.138388e+14,61964.270982,1.315098e+06,0.502997,308.50
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
171554,2021-04-26,RELIANCE,EQ,1904.35,1920.00,1962.00,1911.50,1938.00,1937.85,1941.32,9620785,1.867699e+15,259137.000000,4.276703e+06,0.444500,1988.65
171555,2021-04-27,RELIANCE,EQ,1937.85,1940.00,1997.20,1938.25,1990.00,1988.65,1978.64,9226547,1.825602e+15,291197.000000,3.772144e+06,0.408800,1997.30
171556,2021-04-28,RELIANCE,EQ,1988.65,1997.85,2008.00,1980.15,1993.15,1997.30,1997.60,7902002,1.578508e+15,247331.000000,3.921560e+06,0.496300,2024.05
171557,2021-04-29,RELIANCE,EQ,1997.30,2022.90,2044.50,2007.30,2020.00,2024.05,2024.21,8035915,1.626634e+15,213153.000000,2.834103e+06,0.352700,1994.50


In [39]:
from sklearn.impute import SimpleImputer
imputer=SimpleImputer(strategy="mean")
df_reliance["target"]=imputer.fit_transform(df_reliance[["target"]])
df_reliance

,Date,Symbol,Series,Prev Close,Open,High,Low,Last,Close,VWAP,Volume,Turnover,Trades,Deliverable Volume,%Deliverble,target
166253,2000-01-03,RELIANCE,EQ,233.05,237.50,251.70,237.50,251.70,251.70,249.37,4456424,1.111319e+14,61964.270982,1.315098e+06,0.502997,271.850000
166254,2000-01-04,RELIANCE,EQ,251.70,258.40,271.85,251.30,271.85,271.85,263.52,9487878,2.500222e+14,61964.270982,1.315098e+06,0.502997,282.500000
166255,2000-01-05,RELIANCE,EQ,271.85,256.65,287.90,256.65,286.75,282.50,274.79,26833684,7.373697e+14,61964.270982,1.315098e+06,0.502997,294.350000
166256,2000-01-06,RELIANCE,EQ,282.50,289.00,300.70,289.00,293.50,294.35,295.45,15682286,4.633254e+14,61964.270982,1.315098e+06,0.502997,314.550000
166257,2000-01-07,RELIANCE,EQ,294.35,295.00,317.90,293.00,314.50,314.55,308.91,19870977,6.138388e+14,61964.270982,1.315098e+06,0.502997,308.500000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
171554,2021-04-26,RELIANCE,EQ,1904.35,1920.00,1962.00,1911.50,1938.00,1937.85,1941.32,9620785,1.867699e+15,259137.000000,4.276703e+06,0.444500,1988.650000
171555,2021-04-27,RELIANCE,EQ,1937.85,1940.00,1997.20,1938.25,1990.00,1988.65,1978.64,9226547,1.825602e+15,291197.000000,3.772144e+06,0.408800,1997.300000
171556,2021-04-28,RELIANCE,EQ,1988.65,1997.85,2008.00,1980.15,1993.15,1997.30,1997.60,7902002,1.578508e+15,247331.000000,3.921560e+06,0.496300,2024.050000
171557,2021-04-29,RELIANCE,EQ,1997.30,2022.90,2044.50,2007.30,2020.00,2024.05,2024.21,8035915,1.626634e+15,213153.000000,2.834103e+06,0.352700,1994.500000


In [40]:

df_reliance["Date"]=df_reliance["Date"].sort_values()
df_reliance

,Date,Symbol,Series,Prev Close,Open,High,Low,Last,Close,VWAP,Volume,Turnover,Trades,Deliverable Volume,%Deliverble,target
166253,2000-01-03,RELIANCE,EQ,233.05,237.50,251.70,237.50,251.70,251.70,249.37,4456424,1.111319e+14,61964.270982,1.315098e+06,0.502997,271.850000
166254,2000-01-04,RELIANCE,EQ,251.70,258.40,271.85,251.30,271.85,271.85,263.52,9487878,2.500222e+14,61964.270982,1.315098e+06,0.502997,282.500000
166255,2000-01-05,RELIANCE,EQ,271.85,256.65,287.90,256.65,286.75,282.50,274.79,26833684,7.373697e+14,61964.270982,1.315098e+06,0.502997,294.350000
166256,2000-01-06,RELIANCE,EQ,282.50,289.00,300.70,289.00,293.50,294.35,295.45,15682286,4.633254e+14,61964.270982,1.315098e+06,0.502997,314.550000
166257,2000-01-07,RELIANCE,EQ,294.35,295.00,317.90,293.00,314.50,314.55,308.91,19870977,6.138388e+14,61964.270982,1.315098e+06,0.502997,308.500000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
171554,2021-04-26,RELIANCE,EQ,1904.35,1920.00,1962.00,1911.50,1938.00,1937.85,1941.32,9620785,1.867699e+15,259137.000000,4.276703e+06,0.444500,1988.650000
171555,2021-04-27,RELIANCE,EQ,1937.85,1940.00,1997.20,1938.25,1990.00,1988.65,1978.64,9226547,1.825602e+15,291197.000000,3.772144e+06,0.408800,1997.300000
171556,2021-04-28,RELIANCE,EQ,1988.65,1997.85,2008.00,1980.15,1993.15,1997.30,1997.60,7902002,1.578508e+15,247331.000000,3.921560e+06,0.496300,2024.050000
171557,2021-04-29,RELIANCE,EQ,1997.30,2022.90,2044.50,2007.30,2020.00,2024.05,2024.21,8035915,1.626634e+15,213153.000000,2.834103e+06,0.352700,1994.500000


In [41]:
#price change in that dte
df_reliance["Price Chane"]=df_reliance["Close"]-df_reliance["Open"]


In [42]:
#daily rerturns
df_reliance["Daily Returns"]=((df_reliance["Close"]-df_reliance["Open"])/df_reliance["Open"])*100
#high and low differ
df_reliance["High Low Difference"]=df_reliance["High"]-df_reliance["Low"]

In [43]:
df_reliance

,Date,Symbol,Series,Prev Close,Open,High,Low,Last,Close,VWAP,Volume,Turnover,Trades,Deliverable Volume,%Deliverble,target,Price Chane,Daily Returns,High Low Difference
166253,2000-01-03,RELIANCE,EQ,233.05,237.50,251.70,237.50,251.70,251.70,249.37,4456424,1.111319e+14,61964.270982,1.315098e+06,0.502997,271.850000,14.20,5.978947,14.20
166254,2000-01-04,RELIANCE,EQ,251.70,258.40,271.85,251.30,271.85,271.85,263.52,9487878,2.500222e+14,61964.270982,1.315098e+06,0.502997,282.500000,13.45,5.205108,20.55
166255,2000-01-05,RELIANCE,EQ,271.85,256.65,287.90,256.65,286.75,282.50,274.79,26833684,7.373697e+14,61964.270982,1.315098e+06,0.502997,294.350000,25.85,10.072083,31.25
166256,2000-01-06,RELIANCE,EQ,282.50,289.00,300.70,289.00,293.50,294.35,295.45,15682286,4.633254e+14,61964.270982,1.315098e+06,0.502997,314.550000,5.35,1.851211,11.70
166257,2000-01-07,RELIANCE,EQ,294.35,295.00,317.90,293.00,314.50,314.55,308.91,19870977,6.138388e+14,61964.270982,1.315098e+06,0.502997,308.500000,19.55,6.627119,24.90
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
171554,2021-04-26,RELIANCE,EQ,1904.35,1920.00,1962.00,1911.50,1938.00,1937.85,1941.32,9620785,1.867699e+15,259137.000000,4.276703e+06,0.444500,1988.650000,17.85,0.929687,50.50
171555,2021-04-27,RELIANCE,EQ,1937.85,1940.00,1997.20,1938.25,1990.00,1988.65,1978.64,9226547,1.825602e+15,291197.000000,3.772144e+06,0.408800,1997.300000,48.65,2.507732,58.95
171556,2021-04-28,RELIANCE,EQ,1988.65,1997.85,2008.00,1980.15,1993.15,1997.30,1997.60,7902002,1.578508e+15,247331.000000,3.921560e+06,0.496300,2024.050000,-0.55,-0.027530,27.85
171557,2021-04-29,RELIANCE,EQ,1997.30,2022.90,2044.50,2007.30,2020.00,2024.05,2024.21,8035915,1.626634e+15,213153.000000,2.834103e+06,0.352700,1994.500000,1.15,0.056849,37.20


In [44]:
X = df_reliance[
    [
        "Prev Close",
        "Open",
        "High",
        "Low",
        "VWAP",
        "Volume",
        "Trades",
        "Deliverable Volume",
        "%Deliverble",
        "Price Chane",
        "High Low Difference",
        "Daily Returns"    ]
]


In [45]:
X.shape

(5306, 12)

In [46]:
y = df_reliance["target"]

In [47]:
y.shape

(5306,)

In [48]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,random_state=42,test_size=0.2)


In [49]:
print(X_train.shape)
print(X_test)
X_test

(4244, 12)
        Prev Close     Open     High      Low     VWAP    Volume  \
167064      278.10   276.00   282.75   276.00   280.85   2662922   
170292     1020.50  1021.00  1025.80  1005.45  1015.77   2529074   
171492     2016.40  2021.40  2058.00  2001.40  2036.58  14271669   
169318      730.30   733.30   745.60   733.30   740.81   2692505   
168117     1669.15  1664.00  1697.85  1647.10  1667.95   3453696   
...            ...      ...      ...      ...      ...       ...   
170110      991.70   992.65  1002.70   991.00   996.31   3659045   
169344      685.15   676.90   696.40   675.00   686.29   2363960   
167525      506.70   505.90   509.90   498.45   503.69   8209129   
166973      271.85   272.50   277.90   271.00   274.98   7450726   
170534     1291.20  1290.00  1298.00  1281.50  1287.46   7107670   

               Trades  Deliverable Volume  %Deliverble  Price Chane  \
167064   61964.270982            595100.0       0.2235         5.05   
170292   50367.000000         

,Prev Close,Open,High,Low,VWAP,Volume,Trades,Deliverable Volume,%Deliverble,Price Chane,High Low Difference,Daily Returns
167064,278.10,276.00,282.75,276.00,280.85,2662922,61964.270982,595100.0,0.2235,5.05,6.75,1.829710
170292,1020.50,1021.00,1025.80,1005.45,1015.77,2529074,50367.000000,1293631.0,0.5115,0.75,20.35,0.073457
171492,2016.40,2021.40,2058.00,2001.40,2036.58,14271669,456106.000000,6676332.0,0.4678,33.30,56.60,1.647373
169318,730.30,733.30,745.60,733.30,740.81,2692505,58859.000000,1467444.0,0.5450,10.00,12.30,1.363698
168117,1669.15,1664.00,1697.85,1647.10,1667.95,3453696,61964.270982,1395468.0,0.4041,-6.95,50.75,-0.417668
...,...,...,...,...,...,...,...,...,...,...,...,...
170110,991.70,992.65,1002.70,991.00,996.31,3659045,90238.000000,1969301.0,0.5382,2.35,11.70,0.236740
169344,685.15,676.90,696.40,675.00,686.29,2363960,73241.000000,1232477.0,0.5214,11.65,21.40,1.721081
167525,506.70,505.90,509.90,498.45,503.69,8209129,61964.270982,2916287.0,0.3552,-4.30,11.45,-0.849970
166973,271.85,272.50,277.90,271.00,274.98,7450726,61964.270982,1346768.0,0.1808,0.95,6.90,0.348624


In [50]:
from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

In [51]:
from sklearn.metrics import (
    mean_absolute_error,
    mean_squared_error,
    r2_score
)
mae = mean_absolute_error(y_test, y_pred)

mse = mean_squared_error(y_test, y_pred)

rmse = mse ** 0.5

r2 = r2_score(y_test, y_pred)

print("Mean Absolute Error :", mae)
print("Mean Squared Error :", mse)
print("Root Mean Squared Error :", rmse)
print("R2 Score :", r2)

Mean Absolute Error : 16.368390272766554
Mean Squared Error : 1835.5319964974599
Root Mean Squared Error : 42.84310908999789
R2 Score : 0.9942649198046569


In [52]:
import pandas as pd

results = pd.DataFrame({
    "Actual": y_test,
    "Predicted": y_pred
})

results.head(10)

,Actual,Predicted
167064,282.30,284.013586
170292,1016.00,1016.951482
171492,2099.40,2047.274344
169318,751.55,745.412162
168117,1666.15,1666.930005
167575,556.85,567.896195
167979,1265.75,1263.584560
169403,782.10,782.164444
170794,882.70,900.136620
168056,1315.70,1316.488321


In [53]:
from sklearn.tree import DecisionTreeRegressor

dt = DecisionTreeRegressor(random_state=42)

dt.fit(X_train, y_train)

y_pred_dt = dt.predict(X_test)

In [54]:
from sklearn.ensemble import RandomForestRegressor

rf = RandomForestRegressor(
    n_estimators=100,
    random_state=42
)

rf.fit(X_train, y_train)

y_pred_rf = rf.predict(X_test)

In [55]:
from sklearn.ensemble import GradientBoostingRegressor

gb = GradientBoostingRegressor(random_state=42)

gb.fit(X_train, y_train)

y_pred_gb = gb.predict(X_test)

In [56]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np

def evaluate_model(model_name, y_test, y_pred):
    print(f"\n{model_name}")
    print("MAE :", mean_absolute_error(y_test, y_pred))
    print("RMSE:", np.sqrt(mean_squared_error(y_test, y_pred)))
    print("R2  :", r2_score(y_test, y_pred))

In [57]:
evaluate_model("Linear Regression", y_test, y_pred)
evaluate_model("Decision Tree", y_test, y_pred_dt)
evaluate_model("Random Forest", y_test, y_pred_rf)
evaluate_model("Gradient Boosting", y_test, y_pred_gb)


Linear Regression
MAE : 16.368390272766554
RMSE: 42.84310908999789
R2  : 0.9942649198046569

Decision Tree
MAE : 24.883239171374765
RMSE: 53.52162665478332
R2  : 0.9910497298456321

Random Forest
MAE : 18.293205363504203
RMSE: 45.93179586281779
R2  : 0.9934081944499814

Gradient Boosting
MAE : 17.808735309943984
RMSE: 44.88486011455558
R2  : 0.9937052673743743


Purpose:

To compare the performance of different regression algorithms for predicting the next day's closing price of RELIANCE.

Observation:

Linear Regression achieved the highest R² score (0.9943) and the lowest MAE and RMSE among all tested models. Gradient Boosting and Random Forest also performed well, while Decision Tree produced comparatively larger prediction errors.

Conclusion:

Linear Regression was selected as the final model because it provided the best overall predictive performance on the test dataset.

In [58]:
import joblib 
type(model)
joblib.dump(model,"RELIANCE.pkl")

['RELIANCE.pkl']

In [ ]:
from sklearn.metrics import r2_score

print(r2_score(y_test, y_pred))

0.9943357861477564
